src/python/dxpy/system_requirements.py

```diff
@@ -0,0 +1,182 @@
+
+from collections import defaultdict
+import copy
+
+from .utils import merge
+from .compat import basestring
+from .exceptions import (err_exit, DXError, DXCLIError)
+
+'''
+System Requirements
++++++++++++++++++++
+
+A module containing utility methods useful for packing and unpacking
+system requirements.
+'''
+
+class SystemRequirementsDict(object):
+    def __init__(self, instance_type=None, cluster_spec=None):
+        self.instance_type = instance_type
+        self.cluster_spec = cluster_spec
+
+    @classmethod
+    def entrypoint2instcount(cls, instance_count, entrypoint="*"):
+        """
+        Returns a dictionary {entrypoint: instance_count}. The instance_count should
+        be either a:
+        * string, eg. 6
+        * dictionary, eg. {"main": 4, "other_function": 2}
+        """
+        if isinstance(instance_count, basestring):
+            # By default, all entry points ("*") should use this instance type
+            print("instance_count", instance_count)
+            return {entrypoint: int(instance_count)}
+        elif isinstance(instance_count, dict):
+            # instance_type is a map of entry point to instance count
+            return {k: int(v) for k, v in instance_count.items()}
+        else:
+            raise DXError('Expected instance_count field to be either a string or a dict')
+
+    @classmethod
+    def from_instance_type(cls, instance_type_arg, entrypoint="*"):
+        """
+        Returns SystemRequirementsDict with instance_type that can be passed as a
+        "systemRequirements" input to job/new or run/ API calls. The instance_type_arg
+        should be either a:
+        * string, eg. mem1_ssd1_x2
+        * dictionary, eg. {"main": "mem2_hdd2_x2", "other_function": "mem2_hdd2_x1"}
+        """
+        if instance_type_arg is None:
+            return cls(instance_type=None)
+        elif isinstance(instance_type_arg, basestring):
+            # By default, all entry points ("*") should use this instance type
+            return cls(instance_type={entrypoint: {"instanceType": instance_type_arg}})
+        elif isinstance(instance_type_arg, dict):
+            # instance_type is a map of entry point to instance type
+            return cls(instance_type={fn: {"instanceType": fn_inst} for fn, fn_inst in instance_type_arg.items()})
+        else:
+            raise DXError('Expected instance_type field to be either a string or a dict')
+
+    @classmethod
+    def from_cluster_spec(cls, app_sys_reqs, instance_count_arg):
```

Comment:
I think `app_sys_reqs` and `instance_count_arg` should generate separate SRDs, and `add_override` should be used to merge them

```
srd_default = SRD.from_sys_req(app_sys_reqs["systemRequirements"])
srd_runtime = SRD.from_instance_count(instance_count_arg)
srd_cluster_spec = srd_default.add_override(srd_runtime)
...
srd = srd_cluster_spec + srd_instance_type

--//--

class SRD(object):
    ...
    @classmethod
    def from_cluster_spec(cls, cluster_specs_by_entry): ...
    @classmethod
    def from_sys_req(cls, sys_req):
        # Resolve cluster_specs_by_entry from sys_req (might not be necessary)
        ...
    @classmethod
    def from_instance_count(cls, instance_count_arg, entry="*"):
        # Magic for resolving cluster_specs_by_entry
        ...
    def add_override(self, srd):
        # Magic for overriding self with srd
        ...
```

This way, the nasty logic is encapsulated in `add_override`, and we can test that extensively. Could help with readability too

WDYT?

Reply:
The instance_count is just a number (or dict) passed by the user, so I don't think it's possible to create a valid SRD from it on it's own (the invariant is that SRD.as_dict() can be a "systemRequirements" passed in the /run api call).

Reply:
You are mostly right about the invariant, but it can be interpreted more flexibly. `as_dict` just needs to provide a systemRequirements-like object that can be passed to the API

When instance_count is a number, what do we assume about it? We map it to `{"*": {"clusterSpec": "initialInstanceCount"}}`. We do not need to worry about the rest of `"*.clusterSpec"`

When instance_count is a dict (e.g., `{"*": 2, "foo": 4}`), we map it to `{"*": {"clusterSpec": {"initialInstanceCount": 2}}, "foo": {"clusterSpec": {"initialInstanceCount": 4}}}`

If the instance-count--based SRD was passed directly to the API without downstream changes, then the call would fail on the API side

So this is where `add_override` can handle adding the instance-count--based SRD to the default SRD

--//--

Alternatively you could start with the default SRD, and add a separate method for overriding based on the instance count

```
class SRD(object):
    @classmethod
    def from_sys_req(...): ...
    def override_instance_count(self, instance_count):
        # Nasty logic
```

In this design, the instance count is treated differently, and would never generate an SRD on its own